### ISSUE TITLE
List[int/float/str] query parameter validation not working correctly for non-multi usage

### ISSUE THREAD

**Issue by TheGreatRefrigerator** [tiangolo/fastapi#2500]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, minimal, reproducible, example with my use case:

```Python
from typing import List, Any

from fastapi import FastAPI, Query, Depends

app = FastAPI(
    title="test"
)


def workaround(bbox: List[str] = Query([])) -> List[float]:

    return [float(x) for x in bbox[0].split(',')]


@app.get("/float-list-test", response_model=List[float], summary="Multi query param float list")
def get_float_list(
        bbox: List[float] = Query([])
) -> Any:
    return bbox


@app.get("/workaround", response_model=List[float], summary="Multi query param float list")
def get_float_list(
        bbox: List[float] = Depends(workaround)
) -> Any:
    print(bbox)
    return bbox

```

### Description

* Open the browser and call the endpoint `/float-list-test?bbox=1.2,2.1,3.5,4.3`.
* It returns a JSON with
```
{
"detail":
  [
    {
      "loc":["query","bbox",0],
      "msg":"value is not a valid float",
      "type":"type_error.float"}
  ]
}
```
* But I expected it to return `[1.2,2.1,3.5,4.3]`.
* Working:
- `/float-list-test?bbox=1.2&bbox=2.1&bbox=3.5&bbox=4.3
- `/workaround?bbox=1.2, 3.1, 1.2

The workaround will produce an OpenAPI entry of `array[string]` though which is not correct.

### Environment

* OS: [macOS(10.14.6)]
* FastAPI Version: [0.61.2]
* Python version: [3.9.0]

### Additional context

My ultimate goal is to be able to pass a bbox for a spatial query, which will also be validated and return a proper validation error. Currently using following model:
```
class BBoxModel(BaseModel):
    __root__: List[float]

    def __iter__(self):
        return iter(self.__root__)

    def __getitem__(self, item):
        return self.__root__[item]

    @root_validator(pre=True)
    def validate_len(cls, values):
        if len(values['__root__']) != 4:
            raise ValueError("bbox needs 4 values: west(lon), south(lat), east(lon), north(lat)")
        return values

    @root_validator(pre=True)
    def validate_coords(cls, values):
        if abs(values['__root__'][0]) > 180:
            raise ValueError("invalid west longitude")
        if abs(values['__root__'][1]) > 90:
            raise ValueError("invalid south latitude")
        if abs(values['__root__'][2]) > 180:
            raise ValueError("invalid east longitude")
        if abs(values['__root__'][3]) > 90:
            raise ValueError("invalid north latitude")
        return values
```

But i can't pass this as model for the bbox variable like
```
@app.get("/bbox")
def get_bbox(
        bbox: BBoxModel = Query([])
) 
```
as it can only be used in the body
```
[...]fastapi/dependencies/utils.py", line 331, in get_dependant
    assert isinstance(
AssertionError: Param: bbox can only be a request body, using Body(...)
```

Right now i'm using this dependency check to work around it, but i would like to have a solution more integrated with the powers of fast api and pydantic.
```
def get_valid_bbox(bbox: List[str] = Query(
    ["-180., -90., 180., 90"],
    title="Bbox",
    description="""Bbox in the form of west(lon), south(lat), east(lon), north(lat)"""
    )
):
    if len(bbox) != 4:
        bbox = [float(x) for x in bbox[0].split(',')]
    try:
        bbox_obj = BBoxModel.parse_obj(bbox)
    except ValidationError as e:
        for error in e.errors():
            error['loc'] = ["bbox"]
        raise HTTPException(
            status_code=422,
            detail=e.errors()
        )
    return bbox_obj

# and then the endpoint:
@app.get("/bbox")
def get_bbox(
    bbox: Optional[list] = Depends(deps.get_valid_bbox)
)
```

Sorry for the long story

**Comment by falkben** [comment]
Yes, you will have to write a custom dependency to solve this, as the only supported way to enter a list of query args is how you have it (`bbox=1.2&bbox=2.1&bbox=3.5&bbox=4.3`).

If you wanted some validation (that they are all floats, for instance), you could create a pydantic model and, after you separate the list, create an instance of the model by passing in the parsed list.

**Comment by sm-Fifteen** [comment]
This style is valid in OpenAPI (this is the [non-exploded array style for parameters](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#parameter-object)), but FastAPI only supports the default style, which is `form` for query parameters, meaning `explode` in turn defaults to `true`, as per the spec.

From what I remember from the last time a feature like this was requested, fine control over the data encoding of FastAPI route parameters isn't a priority, since clients are expected to adjust their parameter encoding style in consequence. It's probably possible to pass custom OpenAPI schema information in your `Query` constructor (unrecognized keyword parameters are written as extra properties in the openapi.yml file) and using a custom parsing logic, like with your `workaround` example.
